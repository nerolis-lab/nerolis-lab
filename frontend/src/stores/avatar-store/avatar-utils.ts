import { getPokemon } from 'sleepapi-common'

export function findPokemon(name: string) {
  const trimmedName = name.split('_')
  const pokemonName = trimmedName.filter((word) => !['shiny', 'happy'].includes(word)).join('_')
  return getPokemon(pokemonName)
}
