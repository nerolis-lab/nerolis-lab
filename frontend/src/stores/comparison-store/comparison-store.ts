import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useTeamStore } from '@/stores/team/team-store'
import { defineStore } from 'pinia'
import { DOMAIN_VERSION, type MemberProduction } from 'sleepapi-common'

export interface ComparisonState {
  members: MemberProduction[]
  teamIndex?: number
  timeWindow: '8H' | '24H'
  domainVersion: number
}

const defaultState = (attrs?: Partial<ComparisonState>): ComparisonState => ({
  members: [],
  teamIndex: undefined,
  timeWindow: '24H',
  domainVersion: 0,
  ...attrs
})

const MAX_COMPARISON_MEMBERS = 10

export const useComparisonStore = defineStore('comparison', {
  state: (): ComparisonState => defaultState(),
  getters: {
    getMemberProduction: (state) => (externalId: string) =>
      state.members.find((member) => member.externalId === externalId),
    fullTeam: (state) => state.members.length >= MAX_COMPARISON_MEMBERS,
    currentTeam: (state) => {
      const teamStore = useTeamStore()
      return state.teamIndex != null ? teamStore.teams[state.teamIndex] : undefined
    }
  },
  actions: {
    invalidateCache() {
      if (this.domainVersion !== DOMAIN_VERSION) {
        const updatedKeys = Object.keys(defaultState()) as Array<keyof ComparisonState>
        const cachedKeys = Object.keys(this.$state) as Array<keyof ComparisonState>

        cachedKeys.forEach((key) => {
          if (!updatedKeys.includes(key)) {
            delete this.$state[key]
          }
        })

        this.$patch(defaultState({ domainVersion: DOMAIN_VERSION }))
      }
    },
    addMember(member: MemberProduction) {
      this.members.push(member)
    },
    removeMember(externalId: string, index: number) {
      const pokemonStore = usePokemonStore()
      pokemonStore.removePokemon(externalId, 'compare')
      this.members.splice(index, 1)
    }
  },
  persist: true
})
