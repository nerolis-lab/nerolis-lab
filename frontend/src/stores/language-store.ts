import { defineStore } from 'pinia'
import 'sleepapi-common'

export interface LanguageState {
  currentLanguage: string
  hasManuallyChanged: boolean
}

const availableLanguages = [
  {
    code: 'en',
    name: 'English',
    nativeName: 'English'
  },
  {
    code: 'ja',
    name: 'Japanese',
    nativeName: '日本語'
  }
]

export const useLanguageStore = defineStore('language', {
  state: (): LanguageState => {
    return {
      currentLanguage: 'en', // Default fallback
      hasManuallyChanged: false
    }
  },
  getters: {
    availableLanguages: () => availableLanguages,
    currentLanguageData: (state) => {
      return availableLanguages.find((lang) => lang.code === state.currentLanguage) || availableLanguages[0]
    }
  },
  actions: {
    initializeLanguage() {
      // Only auto-detect if user hasn't manually changed language before
      if (!this.hasManuallyChanged) {
        const browserLanguage = this.detectBrowserLanguage()
        if (browserLanguage && this.isLanguageSupported(browserLanguage)) {
          this.currentLanguage = browserLanguage
        }
      }

      // Update document language attribute for accessibility
      document.documentElement.lang = this.currentLanguage
    },
    detectBrowserLanguage(): string | undefined {
      // Get browser language, strip region (e.g., 'en-US' -> 'en')
      const browserLang = navigator.language?.split('-').at(0)
      return browserLang
    },
    isLanguageSupported(languageCode: string): boolean {
      return this.availableLanguages.some((lang) => lang.code === languageCode)
    },
    changeLanguage(languageCode: string) {
      if (!this.isLanguageSupported(languageCode)) {
        logger.warn(`Language '${languageCode}' is not supported`)
        return
      }

      this.currentLanguage = languageCode
      this.hasManuallyChanged = true

      // Update document language attribute for accessibility
      document.documentElement.lang = languageCode
    }
  },
  persist: true
})
