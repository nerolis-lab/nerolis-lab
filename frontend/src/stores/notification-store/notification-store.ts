import { NotificationService } from '@/services/notification-service/notification-service'
import { defineStore } from 'pinia'
import type { UserNotification } from 'sleepapi-common'

export interface NotificationState {
  unread: boolean
  notifications: UserNotification[]
}

export const useNotificationStore = defineStore('notification', {
  state: (): NotificationState => {
    return {
      unread: false,
      notifications: []
    }
  },
  actions: {
    async sync(): Promise<UserNotification[]> {
      const response = await NotificationService.getNotifications()
      this.notifications = [...response.notifications]
      return this.notifications
    }
  },
  persist: true
})
