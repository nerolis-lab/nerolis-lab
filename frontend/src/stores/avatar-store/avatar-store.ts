import { findPokemon } from '@/stores/avatar-store/avatar-utils'
import { defineStore } from 'pinia'

export interface AvatarState {
  avatars: Record<string, string>
}

export interface Avatar {
  name: string
  path: string
  displayName: string
  pokedexNumber: number
}

export const useAvatarStore = defineStore('avatar', {
  state: (): AvatarState => ({
    avatars: {}
  }),

  getters: {
    getAvatarPath:
      (state) =>
      (avatarName: string): string => {
        const relativePath = state.avatars[avatarName] ?? state.avatars['default']
        return relativePath ? `/images/avatar/${relativePath}` : '/images/avatar/default.png'
      },
    getBasePokemonAvatars: (state) => {
      const result: Avatar[] = [
        { name: 'default', path: state.avatars['default'], displayName: 'Default', pokedexNumber: 0 }
      ]

      const basePokemon = Object.entries(state.avatars)
        .filter(([name, path]) => path.includes('portrait/') && !name.includes('shiny'))
        .map(([name, path]) => {
          const pokemon = findPokemon(name)
          return {
            name,
            path,
            displayName: pokemon.displayName,
            pokedexNumber: pokemon.pokedexNumber
          }
        })
        .sort((a, b) => a.pokedexNumber - b.pokedexNumber)

      return result.concat(basePokemon)
    }
  },
  actions: {
    async loadAvatars(): Promise<void> {
      try {
        logger.info('Repopulating avatars')
        const response = await fetch('/images/avatar/avatars.json')
        const data = await response.json()
        this.avatars = data
      } catch {
        logger.error('Error loading avatars.json')
        this.avatars = { default: 'default.png' }
      }
    }
  },
  persist: true
})
