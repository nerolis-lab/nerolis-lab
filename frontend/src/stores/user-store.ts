import router from '@/router/router'
import { AuthService } from '@/services/login/auth-service'
import { DISCORD_REDIRECT_URI } from '@/services/login/discord-service'
import { UserService } from '@/services/user/user-service'
import { clearCacheAndLogout } from '@/stores/store-service'
import { useTeamStore } from '@/stores/team/team-store'
import { defineStore } from 'pinia'
import {
  AuthProvider,
  EXPERT_ISLANDS,
  ISLANDS,
  MAX_POT_SIZE,
  Roles,
  type AreaInstances,
  type AuthProviders,
  type IslandInstance,
  type IslandShortName,
  type LoginResponse,
  type UserSettingsResponse
} from 'sleepapi-common'

export interface UserState {
  name: string
  avatar: string | null
  externalId: string | null
  friendCode: string | null
  auth: AuthProviders | null
  role: Roles // TODO: make Role[]
  islands: AreaInstances
  potSize: number
  supporterSince: string | null
  randomizeNicknames: boolean
}

export const useUserStore = defineStore('user', {
  state: (): UserState => {
    const allIslands = [...ISLANDS, ...EXPERT_ISLANDS]
    const islandsRecord = Object.fromEntries(
      allIslands.map((island) => [island.shortName, { ...island, areaBonus: 0 }])
    ) as Record<IslandShortName, IslandInstance>

    return {
      name: 'Guest',
      avatar: null,
      externalId: null,
      friendCode: null,
      role: Roles.Default,
      islands: islandsRecord,
      potSize: MAX_POT_SIZE,
      auth: null,
      supporterSince: null,
      randomizeNicknames: true
    }
  },
  getters: {
    // auth
    loggedIn: (state) => state.auth != null,
    isAdmin: (state) => state.role === Roles.Admin,
    isSupporter: (state) => state.role === Roles.Supporter,
    isAdminOrSupporter: (state) => state.role === Roles.Admin || state.role === Roles.Supporter,
    roleData: (state) => {
      return {
        color: state.role === Roles.Admin ? 'primary' : state.role === Roles.Supporter ? 'supporter' : 'surface',
        icon: state.role === Roles.Admin ? 'mdi-crown' : state.role === Roles.Supporter ? 'mdi-heart' : '',
        text: state.role === Roles.Admin ? 'Admin' : state.role === Roles.Supporter ? 'Supporter' : ''
      }
    },
    numberOfLinkedProviders: (state) => {
      return state.auth?.linkedProviders
        ? Object.values(state.auth.linkedProviders).filter((provider) => provider.linked).length
        : 0
    },
    isProviderLinked: (state) => (provider: AuthProvider) => {
      return state.auth?.linkedProviders?.[provider]?.linked === true
    },

    // areas
    islandBonus: (state) => (shortName?: IslandShortName) => {
      const bonus = shortName ? state.islands[shortName].areaBonus : 0
      return 1 + bonus / 100
    },
    baseIslands: (state) => Object.values(state.islands).filter((island) => !island.expert),
    expertIslands: (state) => Object.values(state.islands).filter((island) => island.expert)
  },
  actions: {
    setInitialLoginData(serverData: LoginResponse) {
      this.name = serverData.name
      this.avatar = serverData.avatar ?? 'default'
      this.externalId = serverData.externalId
      this.role = serverData.role
      this.friendCode = serverData.friendCode

      if (this.auth) {
        this.auth = {
          ...this.auth,
          linkedProviders: serverData.auth.linkedProviders
        }
      } else {
        this.auth = serverData.auth
      }
    },
    setUserSettings(userSettings: UserSettingsResponse) {
      this.name = userSettings.name
      this.avatar = userSettings.avatar
      this.role = userSettings.role
      this.potSize = userSettings.potSize
      this.randomizeNicknames = userSettings.randomizeNicknames

      for (const [area, bonus] of Object.entries(userSettings.areaBonuses)) {
        this.islands[area as IslandShortName].areaBonus = bonus
      }

      this.supporterSince = userSettings.supporterSince
    },
    async syncUserSettings() {
      if (this.loggedIn) {
        const userSettings = await UserService.getUserSettings()
        this.setUserSettings(userSettings)
      }
    },
    async login(params: { authCode: string; provider: AuthProvider; originalRoute: string; redirectUri?: string }) {
      const { authCode, provider, originalRoute, redirectUri } = params
      const loginResponse = await AuthService.login(authCode, provider, redirectUri)

      this.setInitialLoginData(loginResponse)

      await this.syncUserSettings()

      const teamStore = useTeamStore()
      await teamStore.syncTeams()

      router.push(originalRoute)
    },
    async unlinkProvider(provider: AuthProvider) {
      await AuthService.unlinkProvider(provider)
      if (this.auth?.activeProvider === provider) {
        this.logout() // will wipe cache and logout
      } else {
        this.auth && (this.auth.linkedProviders[provider].linked = false)
      }
    },
    async refresh() {
      try {
        if (this.auth) {
          const { tokens, activeProvider } = this.auth
          if (Date.now() > tokens.expiryDate) {
            const { refreshToken } = tokens
            const redirectUri = activeProvider === AuthProvider.Discord ? DISCORD_REDIRECT_URI : undefined

            const { access_token, expiry_date } = await AuthService.refresh(refreshToken, activeProvider, redirectUri)
            this.auth.tokens = {
              accessToken: access_token,
              refreshToken,
              expiryDate: expiry_date
            }
          }
        } else {
          this.logout()
        }
      } catch {
        this.logout()
      }
    },
    logout() {
      clearCacheAndLogout()
      router.push('/')
    }
  },
  persist: true
})
