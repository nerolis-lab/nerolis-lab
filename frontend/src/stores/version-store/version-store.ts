import { useAvatarStore } from '@/stores/avatar-store/avatar-store'
import { defineStore } from 'pinia'

export interface VersionState {
  version: string
}

export const useVersionStore = defineStore('version', {
  state: (): VersionState => {
    return {
      version: '1.0.0'
    }
  },
  getters: {
    updateFound: (state) => state.version !== APP_VERSION
  },
  actions: {
    async updateVersion() {
      logger.debug(`Client updating version: ${this.version} -> ${APP_VERSION}`)
      this.version = APP_VERSION

      const avatarStore = useAvatarStore()
      await avatarStore.loadAvatars()
    }
  },
  persist: true
})
