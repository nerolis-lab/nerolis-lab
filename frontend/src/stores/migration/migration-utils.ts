import { useAvatarStore } from '@/stores/avatar-store/avatar-store'
import { useComparisonStore } from '@/stores/comparison-store/comparison-store'
import type { Migration, StoreMap } from '@/stores/migration/migration-type'
import { useNotificationStore } from '@/stores/notification-store/notification-store'
import { usePokedexStore } from '@/stores/pokedex-store/pokedex-store'
import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useTeamStore } from '@/stores/team/team-store'
import { useUserStore } from '@/stores/user-store'
import { useVersionStore } from '@/stores/version-store/version-store'

export function getMigrations(): Migration[] {
  const migrationModules = import.meta.glob<{ default: Migration }>('./migrations/!(*.test).ts', { eager: true })

  const migrations: Migration[] = Object.entries(migrationModules)
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    .map(([_path, module]) => {
      return module.default
    })
    .sort((a, b) => a.version - b.version)

  return migrations
}

export function getStores() {
  const availableStores = {
    avatar: useAvatarStore,
    comparison: useComparisonStore,
    notification: useNotificationStore,
    pokedex: usePokedexStore,
    pokemon: usePokemonStore,
    team: useTeamStore,
    user: useUserStore,
    version: useVersionStore
  }
  const stores: StoreMap = {}
  for (const storeId in availableStores) {
    stores[storeId] = availableStores[storeId as keyof typeof availableStores]()
  }
  return stores
}
