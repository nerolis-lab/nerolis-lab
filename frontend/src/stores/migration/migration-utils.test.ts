import { useAvatarStore } from '@/stores/avatar-store/avatar-store'
import { useComparisonStore } from '@/stores/comparison-store/comparison-store'
import { useNotificationStore } from '@/stores/notification-store/notification-store'
import { usePokedexStore } from '@/stores/pokedex-store/pokedex-store'
import { usePokemonStore } from '@/stores/pokemon/pokemon-store'
import { useTeamStore } from '@/stores/team/team-store'
import { useUserStore } from '@/stores/user-store'
import { useVersionStore } from '@/stores/version-store/version-store'
import { describe, expect, it, vi } from 'vitest'
import { getMigrations, getStores } from './migration-utils'

vi.mock('@/stores/avatar-store/avatar-store')
vi.mock('@/stores/comparison-store/comparison-store')
vi.mock('@/stores/notification-store/notification-store')
vi.mock('@/stores/pokedex-store/pokedex-store')
vi.mock('@/stores/pokemon/pokemon-store')
vi.mock('@/stores/team/team-store')
vi.mock('@/stores/user-store')
vi.mock('@/stores/version-store/version-store')

describe('migration-utils', () => {
  describe('getMigrations', () => {
    it('should return migrations sorted by version', () => {
      const migrations = getMigrations()

      expect(migrations).toHaveLength(2)
      expect(migrations).toMatchSnapshot()
    })
  })

  describe('getStores', () => {
    it('should return all initialized stores', () => {
      const stores = getStores()

      expect(Object.keys(stores)).toEqual([
        'avatar',
        'comparison',
        'notification',
        'pokedex',
        'pokemon',
        'team',
        'user',
        'version'
      ])

      expect(useAvatarStore).toHaveBeenCalled()
      expect(useComparisonStore).toHaveBeenCalled()
      expect(useNotificationStore).toHaveBeenCalled()
      expect(usePokedexStore).toHaveBeenCalled()
      expect(usePokemonStore).toHaveBeenCalled()
      expect(useTeamStore).toHaveBeenCalled()
      expect(useUserStore).toHaveBeenCalled()
      expect(useVersionStore).toHaveBeenCalled()
    })
  })
})
