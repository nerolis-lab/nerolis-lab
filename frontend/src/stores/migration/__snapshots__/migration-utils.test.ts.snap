// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`migration-utils > getMigrations > should return migrations sorted by version 1`] = `
[
  {
    "description": "Sets up initial baseline version.",
    "up": [Function],
    "version": 1,
  },
  {
    "description": "Adds support for multiple providers.",
    "up": [Function],
    "version": 2,
  },
  {
    "description": "Migrates areaBonus Record to islands structure with IslandInstance objects.",
    "up": [Function],
    "version": 3,
  },
]
`;
